More than one native executable file found

### Describe the bug

I have a project developed with spring boot 3.3.x and compiled natively using GraalVM CE 21.0.1+12.1.
When trying to build the container image using kubernetes-maven-plugin version 1.17.0, I get the error “More than one native executable file found in /path-to-my-project/target”.
We can see below the output:
```
mvn k8s:resource k8s:build -DskipTests
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO]
[INFO] 1.0.0-SNAPSHOT													  [pom]
[INFO]  1.0.0-SNAPSHOT  gitsyn                    [pom]
[INFO] Módulo de entrega 1.0.0-SNAPSHOT  gitsyn                [pom]
[INFO] gitsyn-ms                                                          [jar]
[INFO]
[INFO] -------------------< com.tests.gitsyn:gitsyn >--------------------
[INFO] Building 1.0.0-SNAPSHOT [1/4]
[INFO]   from pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn ---
[INFO] k8s: `k8s:resource` goal is skipped.
[INFO]
[INFO] --- k8s:1.17.0:build (default-cli) @ gitsyn ---
[INFO] k8s: `k8s:build` goal is skipped.
[INFO]
[INFO] ----------------< com.tests.gitsyn:gitsyn-config >----------------
[INFO] Building  1.0.0-SNAPSHOT  gitsyn 1.0.0-SNAPSHOT [2/4]
[INFO]   from gitsyn-config/pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn-config ---
[INFO] k8s: `k8s:resource` goal is skipped.
[INFO]
[INFO] --- k8s:1.17.0:build (default-cli) @ gitsyn-config ---
[INFO] k8s: `k8s:build` goal is skipped.
[INFO]
[INFO] ---------------< com.tests.gitsyn:gitsyn-entrega >----------------
[INFO] Building Módulo de entrega 1.0.0-SNAPSHOT  gitsyn 1.0.0-SNAPSHOT [3/4]
[INFO]   from gitsyn-entrega/pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn-entrega ---
[INFO] k8s: `k8s:resource` goal is skipped.
[INFO]
[INFO] --- k8s:1.17.0:build (default-cli) @ gitsyn-entrega ---
[INFO] k8s: `k8s:build` goal is skipped.
[INFO]
[INFO] ------------------< com.tests.gitsyn:gitsyn-ms >------------------
[INFO] Building gitsyn-ms 1.0.0-SNAPSHOT                                  [4/4]
[INFO]   from gitsyn-ms/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn-ms ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for 1.0.0-SNAPSHOT  1.0.0-SNAPSHOT:
[INFO]
[INFO] 1.0.0-SNAPSHOT  {project.aritfactID} SUCCESS [  1.638 s]
[INFO] 1.0.0-SNAPSHOT  gitsyn .... SUCCESS [  0.047 s]
[INFO] 1.0.0-SNAPSHOT  gitsyn  SUCCESS [  0.043 s]
[INFO] gitsyn-ms .......................................... FAILURE [  1.643 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.923 s
[INFO] Finished at: 2024-12-02T18:12:02+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.17.0:resource (default-cli) on project gitsyn-ms: Execution default-cli of goal org.eclipse.jkube:kubernetes-maven-plugin:1.17.0:resource failed: Cannot load service org.eclipse.jkube.springboot.generator.SpringBootGenerator defined in jar:file:/root/.m2/org/eclipse/jkube/jkube-kit-spring-boot/1.17.0/jkube-kit-spring-boot-1.17.0.jar!/META-INF/jkube/generator-default : java.lang.reflect.InvocationTargetException. Aborting: More than one native executable file found in /root/dev/gitsyn/gitsyn-ms/target -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :gitsyn-ms
```

If we have a look at the contents of the target directory, we can see that there is only one executable file, gitsyn-ms-1.0.0-SNAPSHOT:

```
drwxr-xr-x. 5 root root       137 dic  2 18:06 classes
drwxr-xr-x. 3 root root        25 dic  2 18:06 generated-sources
-rwxr-xr-x. 1 root root 176034504 dic  2 18:11 gitsyn-ms-1.0.0-SNAPSHOT
-rw-r--r--. 1 root root  57724840 dic  2 18:08 gitsyn-ms-1.0.0-SNAPSHOT.jar
-rw-r--r--. 1 root root    571205 dic  2 18:08 gitsyn-ms-1.0.0-SNAPSHOT.jar.original
drwxr-xr-x. 3 root root        54 dic  2 18:08 graalvm-reachability-metadata
-rw-r--r--. 1 root root     38488 dic  2 18:11 libawt_headless.so
-rw-r--r--. 1 root root    809536 dic  2 18:11 libawt.so
-rw-r--r--. 1 root root    575400 dic  2 18:11 libawt_xawt.so
-rw-r--r--. 1 root root    239976 dic  2 18:11 libjavajpeg.so
-rwxr-xr-x. 1 root root     15520 dic  2 18:11 libjava.so
-rwxr-xr-x. 1 root root     15520 dic  2 18:11 libjvm.so
-rw-r--r--. 1 root root    588888 dic  2 18:11 liblcms.so
drwxr-xr-x. 2 root root        28 dic  2 18:06 maven-archiver
drwxr-xr-x. 3 root root        35 dic  2 18:06 maven-status
drwxr-xr-x. 3 root root        18 dic  2 18:06 spring-aot
```

If we look at the target directory, we will see that there is an executable named gitsyn-ms-1.0.0-SNAPSHOT (-rwxr-xr-x), which is the correct one, but as the files libjava.so and libjvm.so are also created, which also have execution permissions (-rwxr-xr-xr-x), the error occurs. 
If I change the permissions of libjava.so and libjvm.so to (-rw-r--r--r--.), the error disappears.

It should be possible to tell the plugin which is the name of the executable file and that it does not take into account the others that exist in the same folder.


### Eclipse JKube version

1.17.0

### Component

Kubernetes Maven Plugin

### Apache Maven version

3.8.5

### Gradle version

None

### Steps to reproduce

```
mvn k8s:resource k8s:build -DskipTests
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO]
[INFO] 1.0.0-SNAPSHOT													  [pom]
[INFO]  1.0.0-SNAPSHOT  gitsyn                    [pom]
[INFO] Módulo de entrega 1.0.0-SNAPSHOT  gitsyn                [pom]
[INFO] gitsyn-ms                                                          [jar]
[INFO]
[INFO] -------------------< com.tests.gitsyn:gitsyn >--------------------
[INFO] Building 1.0.0-SNAPSHOT [1/4]
[INFO]   from pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn ---
[INFO] k8s: `k8s:resource` goal is skipped.
[INFO]
[INFO] --- k8s:1.17.0:build (default-cli) @ gitsyn ---
[INFO] k8s: `k8s:build` goal is skipped.
[INFO]
[INFO] ----------------< com.tests.gitsyn:gitsyn-config >----------------
[INFO] Building  1.0.0-SNAPSHOT  gitsyn 1.0.0-SNAPSHOT [2/4]
[INFO]   from gitsyn-config/pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn-config ---
[INFO] k8s: `k8s:resource` goal is skipped.
[INFO]
[INFO] --- k8s:1.17.0:build (default-cli) @ gitsyn-config ---
[INFO] k8s: `k8s:build` goal is skipped.
[INFO]
[INFO] ---------------< com.tests.gitsyn:gitsyn-entrega >----------------
[INFO] Building Módulo de entrega 1.0.0-SNAPSHOT  gitsyn 1.0.0-SNAPSHOT [3/4]
[INFO]   from gitsyn-entrega/pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn-entrega ---
[INFO] k8s: `k8s:resource` goal is skipped.
[INFO]
[INFO] --- k8s:1.17.0:build (default-cli) @ gitsyn-entrega ---
[INFO] k8s: `k8s:build` goal is skipped.
[INFO]
[INFO] ------------------< com.tests.gitsyn:gitsyn-ms >------------------
[INFO] Building gitsyn-ms 1.0.0-SNAPSHOT                                  [4/4]
[INFO]   from gitsyn-ms/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn-ms ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for 1.0.0-SNAPSHOT  1.0.0-SNAPSHOT:
[INFO]
[INFO] 1.0.0-SNAPSHOT  {project.aritfactID} SUCCESS [  1.638 s]
[INFO] 1.0.0-SNAPSHOT  gitsyn .... SUCCESS [  0.047 s]
[INFO] 1.0.0-SNAPSHOT  gitsyn  SUCCESS [  0.043 s]
[INFO] gitsyn-ms .......................................... FAILURE [  1.643 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.923 s
[INFO] Finished at: 2024-12-02T18:12:02+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.17.0:resource (default-cli) on project gitsyn-ms: Execution default-cli of goal org.eclipse.jkube:kubernetes-maven-plugin:1.17.0:resource failed: Cannot load service org.eclipse.jkube.springboot.generator.SpringBootGenerator defined in jar:file:/root/.m2/org/eclipse/jkube/jkube-kit-spring-boot/1.17.0/jkube-kit-spring-boot-1.17.0.jar!/META-INF/jkube/generator-default : java.lang.reflect.InvocationTargetException. Aborting: More than one native executable file found in /root/dev/gitsyn/gitsyn-ms/target -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :gitsyn-ms
```

### Expected behavior

```
[INFO] ------------------< com.tests.gitsyn:gitsyn-ms >------------------
[INFO] Building gitsyn-ms 1.0.0-SNAPSHOT                                  [4/4]
[INFO]   from gitsyn-ms/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- k8s:1.17.0:resource (default-cli) @ gitsyn-ms ---
[INFO] k8s: Generators:
[INFO] k8s:  - dockerfile-simple
[INFO] k8s:  - quarkus
[INFO] k8s:  - spring-boot
[INFO] k8s:  - thorntail-v2
[INFO] k8s:  - wildfly-jar
[INFO] k8s:  - openliberty
[INFO] k8s:  - karaf
[INFO] k8s:  - vertx
[INFO] k8s:  - micronaut
[INFO] k8s:  - helidon
[INFO] k8s:  - java-exec
[INFO] k8s:  - webapp
[INFO] k8s: Generating resources
[INFO] k8s: Enrichers:
[INFO] k8s: - jkube-metadata
[INFO] k8s: - jkube-controller
[INFO] k8s: - jkube-controller-from-configuration
[INFO] k8s: - jkube-service
[INFO] k8s: - jkube-namespace
[INFO] k8s: - jkube-image
[INFO] k8s: - jkube-portname
[INFO] k8s: - jkube-project-label
[INFO] k8s: - jkube-well-known-labels
[INFO] k8s: - jkube-dependency
[INFO] k8s: - jkube-pod-annotations
[INFO] k8s: - jkube-git
[INFO] k8s: - jkube-maven-scm
[INFO] k8s: - jkube-serviceaccount
[INFO] k8s: - jkube-maven-issue-mgmt
[INFO] k8s: - jkube-debug
[INFO] k8s: - jkube-persistentvolumeclaim-storageclass
[INFO] k8s: - jkube-configmap-file
[INFO] k8s: - jkube-secret-file
[INFO] k8s: - jkube-openshift-route
[INFO] k8s: - jkube-openshift-project
[INFO] k8s: - jkube-ingress
[INFO] k8s: - jkube-healthcheck-quarkus
[INFO] k8s: - jkube-healthcheck-spring-boot
[INFO] k8s: - jkube-healthcheck-thorntail-v2
[INFO] k8s: - jkube-healthcheck-wildfly-jar
[INFO] k8s: - jkube-healthcheck-karaf
[INFO] k8s: - jkube-healthcheck-vertx
[INFO] k8s: - jkube-healthcheck-docker
[INFO] k8s: - jkube-healthcheck-webapp
[INFO] k8s: - jkube-healthcheck-micronaut
[INFO] k8s: - jkube-healthcheck-openliberty
[INFO] k8s: - jkube-healthcheck-smallrye
[INFO] k8s: - jkube-healthcheck-helidon
[INFO] k8s: - jkube-prometheus
[INFO] k8s: - jkube-service-discovery
[INFO] k8s: - jkube-container-env-java-options
[INFO] k8s: - jkube-revision-history
[INFO] k8s: - jkube-docker-registry-secret
[INFO] k8s: - jkube-triggers-annotation
[INFO] k8s: - jkube-openshift-imageChangeTrigger
[INFO] k8s: - jkube-name
[INFO] k8s: - jkube-replicas
[INFO] k8s: - jkube-imagepullpolicy
[INFO] k8s: Using resource templates from /root/dev/gitsyn/gitsyn-ms/src/main/jkube
[INFO] k8s: jkube-service: Adding a default service 'gitsyn-ms-service-app' with ports [8080]
[INFO] k8s: jkube-portname: Adding default port name http for port 8080
[INFO] k8s: jkube-debug: Debugging not enabled. To enable try setting the jkube.debug.enabled maven or system property to 'true'
[INFO] k8s: jkube-healthcheck-spring-boot: Adding readiness probe on port 8080, path='/actuator/health', scheme='HTTP', with initial delay 10 seconds
[INFO] k8s: jkube-healthcheck-spring-boot: Adding liveness probe on port 8080, path='/actuator/health', scheme='HTTP', with initial delay 180 seconds
[INFO] k8s: jkube-service-discovery: Using first mentioned service port '8080'
[INFO] k8s: jkube-revision-history: Adding revision history limit to 2
[INFO] k8s: jkube-docker-registry-secret: Adding secrets resources from plugin configuration
[INFO] k8s: Validating resources
[INFO] k8s: validating /root/dev/gitsyn/gitsyn-despliegue/kubernetes/gitsyn-ms-service-app-service.yml resource
[WARNING] Unknown keyword serializer - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword or if it should generate annotations AnnotationKeyword
[WARNING] Unknown keyword deserializer - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword or if it should generate annotations AnnotationKeyword
[INFO] k8s: validating /root/dev/gitsyn/gitsyn-despliegue/kubernetes/gitsyn-ms-deployment.yml resource
[INFO] k8s: validating /root/dev/gitsyn/gitsyn-despliegue/kubernetes/gitsyn-route.yml resource
[INFO]
[INFO] --- k8s:1.17.0:build (default-cli) @ gitsyn-ms ---
[INFO] k8s: Generators:
[INFO] k8s:  - dockerfile-simple
[INFO] k8s:  - quarkus
[INFO] k8s:  - spring-boot
[INFO] k8s:  - thorntail-v2
[INFO] k8s:  - wildfly-jar
[INFO] k8s:  - openliberty
[INFO] k8s:  - karaf
[INFO] k8s:  - vertx
[INFO] k8s:  - micronaut
[INFO] k8s:  - helidon
[INFO] k8s:  - java-exec
[INFO] k8s:  - webapp
JIB> Base image 'harbor.test.local/library/ubi9' does not use a specific image digest - build may not be reproducible
JIB> Containerizing application with the following files:
JIB>    Aplicacion:
JIB>            /root/dev/gitsyn/gitsyn-config/main/resources/dev
JIB>            /root/dev/gitsyn/gitsyn-config/main/resources/dev/gitsyn.properties
JIB>            /root/dev/gitsyn/gitsyn-config/main/resources/dev/logback-spring.xml
JIB>            /root/dev/gitsyn/gitsyn-ms/target/gitsyn-ms-1.0.0-SNAPSHOT
JIB>    Jkube-generated-layer-final-artifact:
JIB>            /root/dev/gitsyn/gitsyn-ms/target/gitsyn-ms-1.0.0-SNAPSHOT.jar
JIB> Getting manifest for base image harbor.test.local/library/ubi9...
JIB> Building aplicacion layer...
JIB> Building jkube-generated-layer-final-artifact layer...
JIB> The base image requires auth. Trying again for harbor.test.local/library/ubi9...
JIB> WWW-Authenticate for harbor.test.local/library/ubi9: Bearer realm="https://harbor.test.local/service/token",service="harbor-registry",scope="repository:library/ubi9:pull"
JIB> bearer auth succeeded for harbor.test.local/library/ubi9
JIB> Using base image with digest: sha256:3942c7ee20130b2f8c1ca153d0c380af185e145229d5e3a8d6bd34d4cde9ab43
JIB> Building jkube-generated-layer-final-artifact layer built sha256:013347b9d49dfc3942cf44fc3f925969ed36ca98908333985c4308b2da23cce0
JIB> Building aplicacion layer built sha256:135871585d9dcfa16a27733f27f0dbb772b3f07167e6d24b82638bb710495fb1
JIB>
JIB> Container entrypoint set to [/opt/apps/gitsyn-ms-1.0.0-SNAPSHOT, -Dapps_config_path=/opt/apps/apps-config, -Dapps_temp_path=/opt/apps/apps-temp, -Dapps_env=dev]
JIB> Building image to tar file...
JIB> [========================      ] 80,0% complete > writing to tar file
JIB> [==============================] 100,0% complete
[INFO] k8s:  /root/dev/gitsyn/gitsyn-ms/target/docker/harbor.test.local/aac/gitsyn/1.0.0-SNAPSHOT/tmp/jib-image.linux-amd64.tar successfully built
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for 1.0.0-SNAPSHOT {project.aritfactID} 1.0.0-SNAPSHOT:
[INFO]
[INFO] 1.0.0-SNAPSHOT                                       SUCCESS [  1.357 s]
[INFO] 1.0.0-SNAPSHOT gitsyn ....                           SUCCESS [  0.048 s]
[INFO] 1.0.0-SNAPSHOT gitsyn                                SUCCESS [  0.043 s]
[INFO] gitsyn-ms .......................................... SUCCESS [ 19.890 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  21.884 s
[INFO] Finished at: 2024-12-02T18:38:19+01:00
[INFO] ------------------------------------------------------------------------
```

### Runtime

OpenShift

### Kubernetes API Server version

1.25.3

### Environment

Linux

### Eclipse JKube Logs

_No response_

### Sample Reproducer Project

_No response_

### Additional context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -27,6 +27,7 @@ Usage:
 * Fix #2667: Add new helm test goal task (`k8s:helm-test` for maven and `k8sHelmTest` for gradle)
 * Fix #3326: Micronaut healthcheck enricher infers overridden server port in application.properties
 * Fix #3354: Build fails with `imageStream` for `buildRecreate` value
+* Fix #3578: Skip executable libs when searching Spring Boot native binary
 * Fix #3623: Updated jkube-images to 0.0.25
 
 ### 1.17.0 (2024-08-13)

--- a/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SpringBootUtil.java
+++ b/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SpringBootUtil.java
@@ -16,10 +16,13 @@ package org.eclipse.jkube.kit.common.util;
 import java.io.File;
 import java.net.URL;
 import java.net.URLClassLoader;
+import java.util.Arrays;
 import java.util.Collections;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Properties;
+import java.util.stream.Collectors;
 
 import org.eclipse.jkube.kit.common.JavaProject;
 import org.eclipse.jkube.kit.common.Plugin;
@@ -32,133 +35,143 @@ import static org.eclipse.jkube.kit.common.util.PropertiesUtil.getPropertiesFrom
  */
 public class SpringBootUtil {
 
-    public static final String SPRING_BOOT_GROUP_ID = "org.springframework.boot";
-//    public static final String SPRING_BOOT_ARTIFACT_ID = "spring-boot";
-    public static final String SPRING_BOOT_DEVTOOLS_ARTIFACT_ID = "spring-boot-devtools";
-    public static final String SPRING_BOOT_MAVEN_PLUGIN_ARTIFACT_ID = "spring-boot-maven-plugin";
-    public static final String SPRING_BOOT_GRADLE_PLUGIN_ARTIFACT_ID = "org.springframework.boot.gradle.plugin";
-    public static final String DEV_TOOLS_REMOTE_SECRET = "spring.devtools.remote.secret";
-    public static final String DEV_TOOLS_REMOTE_SECRET_ENV = "SPRING_DEVTOOLS_REMOTE_SECRET";
+  public static final String SPRING_BOOT_GROUP_ID = "org.springframework.boot";
+  public static final String SPRING_BOOT_DEVTOOLS_ARTIFACT_ID = "spring-boot-devtools";
+  public static final String SPRING_BOOT_MAVEN_PLUGIN_ARTIFACT_ID = "spring-boot-maven-plugin";
+  public static final String SPRING_BOOT_GRADLE_PLUGIN_ARTIFACT_ID = "org.springframework.boot.gradle.plugin";
+  public static final String DEV_TOOLS_REMOTE_SECRET = "spring.devtools.remote.secret";
+  public static final String DEV_TOOLS_REMOTE_SECRET_ENV = "SPRING_DEVTOOLS_REMOTE_SECRET";
 
-    private static final String SPRING_WEB_FLUX_ARTIFACT_ID = "spring-boot-starter-webflux";
-    private static final String PLACEHOLDER_PREFIX = "${";
-    private static final String PLACEHOLDER_SUFFIX = "}";
-    private static final String VALUE_SEPARATOR = ":";
+  private static final String SPRING_WEB_FLUX_ARTIFACT_ID = "spring-boot-starter-webflux";
+  private static final String PLACEHOLDER_PREFIX = "${";
+  private static final String PLACEHOLDER_SUFFIX = "}";
+  private static final String VALUE_SEPARATOR = ":";
 
-    private SpringBootUtil() {}
+  private SpringBootUtil() {}
 
-    /**
-     * Returns the spring boot configuration (supports `application.properties` and `application.yml`)
-     * or an empty properties object if not found, it assumes first profile as default profile.
-     *
-     * @param compileClassLoader compile class loader
-     * @return properties object
-     */
-    public static Properties getSpringBootApplicationProperties(URLClassLoader compileClassLoader) {
-        return getSpringBootApplicationProperties(null, compileClassLoader);
+  /**
+   * Returns the spring boot configuration (supports `application.properties` and `application.yml`)
+   * or an empty properties object if not found, it assumes first profile as default profile.
+   *
+   * @param compileClassLoader compile class loader
+   * @return properties object
+   */
+  public static Properties getSpringBootApplicationProperties(URLClassLoader compileClassLoader) {
+    return getSpringBootApplicationProperties(null, compileClassLoader);
+  }
+
+  /**
+   * Returns the spring boot configuration (supports `application.properties` and `application.yml`)
+   * or an empty properties object if not found
+   *
+   * @param springActiveProfile currently active spring-boot profile
+   * @param compileClassLoader compile class loader
+   * @return properties object
+   */
+  public static Properties getSpringBootApplicationProperties(String springActiveProfile, URLClassLoader compileClassLoader) {
+    URL ymlResource = compileClassLoader.findResource("application.yml");
+    URL propertiesResource = compileClassLoader.findResource("application.properties");
+
+    Properties props = YamlUtil.getPropertiesFromYamlResource(springActiveProfile, ymlResource);
+    props.putAll(getPropertiesFromResource(propertiesResource));
+    if (ymlResource != null) {
+      props.put(JKUBE_INTERNAL_APP_CONFIG_FILE_LOCATION, ymlResource.toString());
+    } else if (propertiesResource != null) {
+      props.put(JKUBE_INTERNAL_APP_CONFIG_FILE_LOCATION, propertiesResource.toString());
     }
+    return new SpringBootPropertyPlaceholderHelper(PLACEHOLDER_PREFIX, PLACEHOLDER_SUFFIX, VALUE_SEPARATOR, true)
+      .replaceAllPlaceholders(props);
+  }
 
-    /**
-     * Returns the spring boot configuration (supports `application.properties` and `application.yml`)
-     * or an empty properties object if not found
-     *
-     * @param springActiveProfile currently active spring-boot profile
-     * @param compileClassLoader compile class loader
-     * @return properties object
-     */
-    public static Properties getSpringBootApplicationProperties(String springActiveProfile, URLClassLoader compileClassLoader) {
-        URL ymlResource = compileClassLoader.findResource("application.yml");
-        URL propertiesResource = compileClassLoader.findResource("application.properties");
+  /**
+   * Determine the spring-boot major version for the current project
+   *
+   * @param javaProject  project
+   * @return spring boot version or null
+   */
+  public static Optional<String> getSpringBootVersion(JavaProject javaProject) {
+    return Optional.ofNullable(JKubeProjectUtil.getAnyDependencyVersionWithGroupId(javaProject, SPRING_BOOT_GROUP_ID));
+  }
 
-        Properties props = YamlUtil.getPropertiesFromYamlResource(springActiveProfile, ymlResource);
-        props.putAll(getPropertiesFromResource(propertiesResource));
-        if (ymlResource != null) {
-            props.put(JKUBE_INTERNAL_APP_CONFIG_FILE_LOCATION, ymlResource.toString());
-        } else if (propertiesResource != null) {
-            props.put(JKUBE_INTERNAL_APP_CONFIG_FILE_LOCATION, propertiesResource.toString());
-        }
-        return new SpringBootPropertyPlaceholderHelper(PLACEHOLDER_PREFIX, PLACEHOLDER_SUFFIX, VALUE_SEPARATOR, true)
-            .replaceAllPlaceholders(props);
+  /**
+   * Returns the currently active spring-boot profile or null if not found.
+   * @param project the JavaProject for which to search the active profile.
+   * @return the currently active spring-boot profile or null if not found.
+   */
+  public static String getSpringBootActiveProfile(JavaProject project) {
+    if (project != null && project.getProperties() != null
+      && project.getProperties().get("spring.profiles.active") != null) {
+      return project.getProperties().get("spring.profiles.active").toString();
     }
+    return null;
+  }
 
-    /**
-     * Determine the spring-boot major version for the current project
-     *
-     * @param javaProject  project
-     * @return spring boot version or null
-     */
-    public static Optional<String> getSpringBootVersion(JavaProject javaProject) {
-        return Optional.ofNullable(JKubeProjectUtil.getAnyDependencyVersionWithGroupId(javaProject, SPRING_BOOT_GROUP_ID));
+  /**
+   * Returns a Map containing the Spring Boot configuration for the applicable plugin (Maven or Gradle).
+   * @param javaProject the JavaProject for which to search the Spring Boot plugin configuration.
+   * @return a Map containing the Spring Boot configuration or an empty Map if no plugin is found.
+   */
+  public static Map<String, Object> getSpringBootPluginConfiguration(JavaProject javaProject) {
+    Plugin mavenPlugin = JKubeProjectUtil.getPlugin(javaProject, SPRING_BOOT_MAVEN_PLUGIN_ARTIFACT_ID);
+    if (mavenPlugin != null) {
+      return mavenPlugin.getConfiguration();
     }
+    Plugin gradlePlugin = JKubeProjectUtil.getPlugin(javaProject, SPRING_BOOT_GRADLE_PLUGIN_ARTIFACT_ID);
+    if (gradlePlugin != null) {
+      return gradlePlugin.getConfiguration();
+    }
+    return Collections.emptyMap();
+  }
 
-    /**
-     * Returns the currently active spring-boot profile or null if not found.
-     * @param project the JavaProject for which to search the active profile.
-     * @return the currently active spring-boot profile or null if not found.
-     */
-    public static String getSpringBootActiveProfile(JavaProject project) {
-        if (project != null && project.getProperties() != null
-              && project.getProperties().get("spring.profiles.active") != null) {
-            return project.getProperties().get("spring.profiles.active").toString();
-        }
-        return null;
-    }
+  public static boolean isSpringBootRepackage(JavaProject project) {
+    Plugin plugin = JKubeProjectUtil.getPlugin(project, SPRING_BOOT_MAVEN_PLUGIN_ARTIFACT_ID);
+    return Optional.ofNullable(plugin)
+      .map(Plugin::getExecutions)
+      .map(e -> e.contains("repackage"))
+      .orElse(false);
+  }
 
-    /**
-     * Returns a Map containing the Spring Boot configuration for the applicable plugin (Maven or Gradle).
-     * @param javaProject the JavaProject for which to search the Spring Boot plugin configuration.
-     * @return a Map containing the Spring Boot configuration or an empty Map if no plugin is found.
-     */
-    public static Map<String, Object> getSpringBootPluginConfiguration(JavaProject javaProject) {
-        Plugin mavenPlugin = JKubeProjectUtil.getPlugin(javaProject, SPRING_BOOT_MAVEN_PLUGIN_ARTIFACT_ID);
-        if (mavenPlugin != null) {
-            return mavenPlugin.getConfiguration();
-        }
-        Plugin gradlePlugin = JKubeProjectUtil.getPlugin(javaProject, SPRING_BOOT_GRADLE_PLUGIN_ARTIFACT_ID);
-        if (gradlePlugin != null) {
-            return gradlePlugin.getConfiguration();
-        }
-        return Collections.emptyMap();
+  public static Plugin getNativePlugin(JavaProject project) {
+    Plugin plugin = JKubeProjectUtil.getPlugin(project, "org.graalvm.buildtools", "native-maven-plugin");
+    if (plugin != null) {
+      return plugin;
     }
+    return JKubeProjectUtil.getPlugin(project, "org.graalvm.buildtools.native", "org.graalvm.buildtools.native.gradle.plugin");
+  }
 
-    public static boolean isSpringBootRepackage(JavaProject project) {
-        Plugin plugin = JKubeProjectUtil.getPlugin(project, SPRING_BOOT_MAVEN_PLUGIN_ARTIFACT_ID);
-        return Optional.ofNullable(plugin)
-            .map(Plugin::getExecutions)
-            .map(e -> e.contains("repackage"))
-            .orElse(false);
+  /**
+   * Returns the native executable artifact produced file by the Spring Boot build or null if not found.
+   * @param project the JavaProject for which to search the native executable artifact
+   * @return the native executable artifact produced file by the Spring Boot build or null if not found
+   * @throws IllegalStateException if more than one native executable artifact is found
+   */
+  public static File findNativeArtifactFile(JavaProject project) {
+    File[] nativeExecutableArtifacts = null;
+    for (String location : new String[] {"", "native/nativeCompile/"}) {
+      nativeExecutableArtifacts = new File(project.getBuildDirectory(), location)
+        .listFiles(f -> f.isFile() && f.canExecute());
+      if (nativeExecutableArtifacts != null && nativeExecutableArtifacts.length > 0) {
+        break;
+      }
     }
+    if (nativeExecutableArtifacts == null) {
+      return null;
+    }
+    if (nativeExecutableArtifacts.length == 1) {
+      return nativeExecutableArtifacts[0];
+    }
+    // TODO: find better way to distinguish between native executables
+    final List<File> filteredBinaries = Arrays.stream(nativeExecutableArtifacts)
+      .filter(f -> !f.getName().endsWith(".so"))
+      .collect(Collectors.toList());
+    if (filteredBinaries.size() == 1) {
+      return filteredBinaries.iterator().next();
+    }
+    throw new IllegalStateException("More than one native executable file found in " + project.getBuildDirectory().getAbsolutePath());
+  }
 
-    public static Plugin getNativePlugin(JavaProject project) {
-        Plugin plugin = JKubeProjectUtil.getPlugin(project, "org.graalvm.buildtools", "native-maven-plugin");
-        if (plugin != null) {
-            return plugin;
-        }
-        return JKubeProjectUtil.getPlugin(project, "org.graalvm.buildtools.native", "org.graalvm.buildtools.native.gradle.plugin");
-    }
-
-    /**
-     * Returns the native executable artifact produced file by the Spring Boot build or null if not found.
-     * @param project the JavaProject for which to search the native executable artifact
-     * @return the native executable artifact produced file by the Spring Boot build or null if not found
-     * @throws IllegalStateException if more than one native executable artifact is found
-     */
-    public static File findNativeArtifactFile(JavaProject project) {
-        for (String location : new String[] {"", "native/nativeCompile/"}) {
-            File nativeArtifactDir = new File(project.getBuildDirectory(), location);
-            File[] nativeExecutableArtifacts = nativeArtifactDir.listFiles(f -> f.isFile() && f.canExecute());
-            if (nativeExecutableArtifacts != null && nativeExecutableArtifacts.length > 0) {
-                if (nativeExecutableArtifacts.length == 1) {
-                    return nativeExecutableArtifacts[0];
-                }
-                throw new IllegalStateException("More than one native executable file found in " + nativeArtifactDir.getAbsolutePath());
-            }
-        }
-        return null;
-    }
-
-    public static boolean hasSpringWebFluxDependency(JavaProject javaProject) {
-        return JKubeProjectUtil.hasDependency(javaProject, SPRING_BOOT_GROUP_ID, SPRING_WEB_FLUX_ARTIFACT_ID);
-    }
+  public static boolean hasSpringWebFluxDependency(JavaProject javaProject) {
+    return JKubeProjectUtil.hasDependency(javaProject, SPRING_BOOT_GROUP_ID, SPRING_WEB_FLUX_ARTIFACT_ID);
+  }
 }
 

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootUtilTest.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootUtilTest.java
@@ -17,15 +17,17 @@ import org.apache.commons.io.FileUtils;
 import org.eclipse.jkube.kit.common.Dependency;
 import org.eclipse.jkube.kit.common.JavaProject;
 import org.eclipse.jkube.kit.common.Plugin;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
-import org.junit.jupiter.api.BeforeEach;
 
 import java.io.File;
 import java.io.IOException;
 import java.net.URL;
 import java.net.URLClassLoader;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.Map;
@@ -33,393 +35,382 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.Properties;
 
-import static org.assertj.core.api.Assertions.assertThatIllegalStateException;
-import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalStateException;
 import static org.assertj.core.api.Assertions.entry;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 
 class SpringBootUtilTest {
 
-    private JavaProject mavenProject;
+  private JavaProject mavenProject;
+
+  @BeforeEach
+  public void setUp() {
+    mavenProject = JavaProject.builder()
+      .properties(new Properties())
+      .build();
+  }
+
+  @Test
+  void getSpringBootApplicationPropertiesLoadsStandardProperties(@TempDir File tempDir) throws IOException {
+    //Given
+    URLClassLoader cl = createClassLoader(tempDir, "/util/springboot/spring-boot-application.properties");
+    //When
+    Properties result = SpringBootUtil.getSpringBootApplicationProperties(cl);
+    //Then
+    assertThat(result).containsOnly(
+      entry("jkube.internal.application-config-file.path", tempDir.toPath().resolve("target/classes/application.properties").toUri().toURL().toString()),
+      entry("spring.application.name", "demoservice"),
+      entry("server.port", "9090")
+    );
+  }
+
+  @Test
+  void getSpringBootVersion() {
+    //Given
+    mavenProject = mavenProject.toBuilder()
+      .dependency(Dependency.builder().groupId("org.springframework.boot").version("1.6.3").build())
+      .build();
+
+    //when
+    Optional<String> result = SpringBootUtil.getSpringBootVersion(mavenProject);
+
+    //Then
+    assertThat(result).isPresent().contains("1.6.3");
+  }
+
+  @Nested
+  class GetSpringBootActiveProfile {
+    @Test
+    void whenNotNull() {
+      //Given
+      mavenProject.getProperties().put("spring.profiles.active", "spring-boot");
+
+      // When
+      String result = SpringBootUtil.getSpringBootActiveProfile(mavenProject);
+
+      //Then
+      assertThat(result).isEqualTo("spring-boot");
+    }
+
+    @Test
+    void whenNull() {
+      assertThat(SpringBootUtil.getSpringBootActiveProfile(null)).isNull();
+    }
+  }
+
+  @Test
+  void getSpringBootApplicationProperties_withCompileClassloader_shouldLoadProperties() {
+    // Given
+    JavaProject javaProject = JavaProject.builder()
+      .compileClassPathElement(Objects.requireNonNull(getClass().getResource("/util/springboot/resources")).getPath())
+      .outputDirectory(new File("target"))
+      .build();
+    URLClassLoader compileClassLoader = JKubeProjectUtil.getClassLoader(javaProject);
+    // When
+    Properties properties = SpringBootUtil.getSpringBootApplicationProperties(compileClassLoader);
+    // Then
+    assertThat(properties)
+      .containsEntry("server.port", "8081");
+  }
+
+  @Nested
+  class GetPropertiesFromYamlResource {
+    @Test
+    void yamlToPropertiesParsing() {
+      assertThat(YamlUtil.getPropertiesFromYamlResource(getClass().getResource("/util/springboot/test-application.yml")))
+        .isNotEmpty()
+        .contains(
+          entry("management.port", "8081"),
+          entry("spring.datasource.url", "jdbc:mysql://127.0.0.1:3306"),
+          entry("example.nested.items[0].value", "value0"),
+          entry("example.nested.items[1].value", "value1"),
+          entry("example.nested.items[2].elements[0].element[0].subelement", "sub0"),
+          entry("example.nested.items[2].elements[0].element[1].subelement", "sub1"),
+          entry("example.1", "integerKeyElement"));
+    }
+
+    @Test
+    void invalidFileThrowsException() {
+      URL resource = SpringBootUtil.class.getResource("/util/springboot/invalid-application.yml");
+      assertThrows(IllegalStateException.class, () -> YamlUtil.getPropertiesFromYamlResource(resource));
+    }
+
+    @Test
+    void nonExistentYamlToProperties() {
+      Properties props = YamlUtil.getPropertiesFromYamlResource(getClass().getResource("/this-file-does-not-exist"));
+      assertThat(props).isNotNull().isEmpty();
+    }
+  }
+
+  @Nested
+  class GetSpringBootApplicationProperties {
+    @Test
+    void withMultipleProfilesAndDefault(@TempDir File tempDir) throws IOException {
+      URLClassLoader cl = createClassLoader(tempDir, "/util/springboot/test-application-with-multiple-profiles.yml");
+      Properties props = SpringBootUtil.getSpringBootApplicationProperties(cl);
+      assertThat(props).isNotEmpty()
+        .contains(
+          entry("spring.application.name", "spring-boot-k8-recipes"),
+          entry("management.endpoints.enabled-by-default", "false"),
+          entry("management.endpoint.health.enabled", "true"))
+        .doesNotContainEntry("cloud.kubernetes.reload.enabled", null);
+    }
+
+    @Test
+    void withMultipleProfilesAndSpecificProfile(@TempDir File tempDir) throws IOException {
+      URLClassLoader cl = createClassLoader(tempDir, "/util/springboot/test-application-with-multiple-profiles.yml");
+      Properties props = SpringBootUtil.getSpringBootApplicationProperties("kubernetes", cl);
+      assertThat(props)
+        .containsEntry("cloud.kubernetes.reload.enabled", "true")
+        .doesNotContain(
+          entry("cloud.kubernetes.reload.enabled", null),
+          entry("spring.application.name", null));
+    }
+  }
+
+  @Nested
+  class GetSpringBootPluginConfiguration {
+    @Test
+    void whenNothingPresent_thenReturnsEmptyMap() {
+      // Given
+      JavaProject javaProject = JavaProject.builder().build();
+      // When
+      Map<String, Object> configuration = SpringBootUtil.getSpringBootPluginConfiguration(javaProject);
+      // Then
+      assertThat(configuration).isEmpty();
+    }
+
+    @Test
+    void whenSpringBootMavenPluginPresent_thenReturnsPluginConfiguration() {
+      // Given
+      JavaProject javaProject = JavaProject.builder()
+        .plugin(Plugin.builder()
+          .groupId("org.springframework.boot")
+          .artifactId("spring-boot-maven-plugin")
+          .configuration(Collections.singletonMap("layout", "ZIP"))
+          .build())
+        .build();
+      // When
+      Map<String, Object> configuration = SpringBootUtil.getSpringBootPluginConfiguration(javaProject);
+      // Then
+      assertThat(configuration).isNotNull().containsEntry("layout", "ZIP");
+    }
+
+    @Test
+    void whenSpringBootGradlePluginPresent_thenReturnsPluginConfiguration() {
+      // Given
+      JavaProject javaProject = JavaProject.builder()
+        .plugin(Plugin.builder()
+          .groupId("org.springframework.boot")
+          .artifactId("org.springframework.boot.gradle.plugin")
+          .configuration(Collections.singletonMap("mainClass", "com.example.ExampleApplication"))
+          .build())
+        .build();
+      // When
+      Map<String, Object> configuration = SpringBootUtil.getSpringBootPluginConfiguration(javaProject);
+      // Then
+      assertThat(configuration).isNotNull().containsEntry("mainClass", "com.example.ExampleApplication");
+    }
+  }
+
+  @Nested
+  class IsSpringBootRepackage {
+    @Test
+    void whenPluginHasRepackageExecution_thenReturnTrue() {
+      // Given
+      JavaProject javaProject = JavaProject.builder()
+        .plugin(Plugin.builder()
+          .groupId("org.springframework.boot")
+          .artifactId("spring-boot-maven-plugin")
+          .executions(Collections.singletonList("repackage"))
+          .build())
+        .build();
+      // When
+      boolean result = SpringBootUtil.isSpringBootRepackage(javaProject);
+      // Then
+      assertThat(result).isTrue();
+    }
+
+    @Test
+    void whenNoExecution_thenReturnFalse() {
+      // Given
+      JavaProject javaProject = JavaProject.builder()
+        .plugin(Plugin.builder()
+          .groupId("org.springframework.boot")
+          .artifactId("spring-boot-maven-plugin")
+          .executions(Collections.emptyList())
+          .build())
+        .build();
+      // When
+      boolean result = SpringBootUtil.isSpringBootRepackage(javaProject);
+      // Then
+      assertThat(result).isFalse();
+    }
+  }
+
+  @Nested
+  class GetNativePlugin {
+    @Test
+    void whenNoNativePluginPresent_thenReturnNull() {
+      assertThat(SpringBootUtil.getNativePlugin(JavaProject.builder().build())).isNull();
+    }
+
+    @Test
+    void whenMavenNativePluginPresent_thenReturnPlugin() {
+      // Given
+      JavaProject javaProject = JavaProject.builder()
+        .plugin(Plugin.builder()
+          .groupId("org.graalvm.buildtools")
+          .artifactId("native-maven-plugin")
+          .build())
+        .build();
+      // When
+      Plugin plugin = SpringBootUtil.getNativePlugin(javaProject);
+      // Then
+      assertThat(plugin).isNotNull();
+    }
+
+    @Test
+    void whenGradleNativePluginPresent_thenReturnPlugin() {
+      // Given
+      JavaProject javaProject = JavaProject.builder()
+        .plugin(Plugin.builder()
+          .groupId("org.graalvm.buildtools.native")
+          .artifactId("org.graalvm.buildtools.native.gradle.plugin")
+          .build())
+        .build();
+      // When
+      Plugin plugin = SpringBootUtil.getNativePlugin(javaProject);
+      // Then
+      assertThat(plugin).isNotNull();
+    }
+  }
+
+  @Nested
+  class FindNativeArtifactFile {
+
+    @TempDir
+    private Path tempDir;
+
+    private JavaProject javaProject;
 
     @BeforeEach
-    public void setUp() {
-        mavenProject = JavaProject.builder()
-          .properties(new Properties())
-          .build();
+    void setUp() {
+      javaProject = JavaProject.builder()
+        .artifactId("sample")
+        .buildDirectory(tempDir.toFile())
+        .plugin(Plugin.builder()
+          .groupId("org.graalvm.buildtools")
+          .artifactId("native-maven-plugin")
+          .build())
+        .build();
     }
 
     @Test
-    void getSpringBootApplicationPropertiesLoadsStandardProperties(@TempDir File tempDir) throws IOException {
-        //Given
-        URLClassLoader cl = createClassLoader(tempDir, "/util/springboot/spring-boot-application.properties");
-        //When
-        Properties result =  SpringBootUtil.getSpringBootApplicationProperties(cl);
-        //Then
-        assertThat(result).containsOnly(
-                entry("jkube.internal.application-config-file.path", tempDir.toPath().resolve("target/classes/application.properties").toUri().toURL().toString()),
-                entry("spring.application.name", "demoservice"),
-                entry("server.port", "9090")
-        );
+    void whenNativeExecutableNotFound_thenReturnNull() {
+      // When
+      File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
+      // Then
+      assertThat(nativeArtifactFound).isNull();
     }
 
     @Test
-    void getSpringBootVersion() {
-        //Given
-        mavenProject = mavenProject.toBuilder()
-          .dependency(Dependency.builder().groupId("org.springframework.boot").version("1.6.3").build())
-          .build();
-
-        //when
-        Optional<String> result = SpringBootUtil.getSpringBootVersion(mavenProject);
-
-        //Then
-        assertThat(result).isPresent().contains("1.6.3");
+    void whenNativeExecutableInStandardMavenBuildDirectory_thenReturnNativeArtifact() throws IOException {
+      // Given
+      File nativeArtifactFile = Files.createFile(tempDir.resolve("sample")).toFile();
+      assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
+      // When
+      File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
+      // Then
+      assertThat(nativeArtifactFound).hasName("sample");
     }
 
     @Test
-    void getSpringBootActiveProfileWhenNotNull() {
-        //Given
-        mavenProject.getProperties().put("spring.profiles.active","spring-boot");
-
-        // When
-        String result = SpringBootUtil.getSpringBootActiveProfile(mavenProject);
-
-        //Then
-        assertThat(result).isEqualTo("spring-boot");
+    void whenNativeExecutableInStandardMavenBuildDirectoryAndImageNameOverridden_thenReturnNativeArtifact() throws IOException {
+      // Given
+      File nativeArtifactFile = Files.createFile(tempDir.resolve("custom-native-name")).toFile();
+      assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
+      // When
+      File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
+      // Then
+      assertThat(nativeArtifactFound).hasName("custom-native-name");
     }
 
     @Test
-    void testGetSpringBootActiveProfileWhenNull() {
-        assertThat(SpringBootUtil.getSpringBootActiveProfile(null)).isNull();
+    void whenNativeExecutableInStandardGradleNativeDirectory_thenReturnNativeArtifact() throws IOException {
+      // Given
+      Files.createDirectories(tempDir.resolve("native").resolve("nativeCompile"));
+      File nativeArtifactFile = Files.createFile(tempDir.resolve("native").resolve("nativeCompile").resolve("sample")).toFile();
+      assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
+      // When
+      File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
+      // Then
+      assertThat(nativeArtifactFound).hasName("sample");
     }
 
     @Test
-    void getSpringBootApplicationProperties_withCompileClassloader_shouldLoadProperties() {
-        // Given
-        JavaProject javaProject = JavaProject.builder()
-            .compileClassPathElement(Objects.requireNonNull(getClass().getResource("/util/springboot/resources")).getPath())
-            .outputDirectory(new File("target"))
-            .build();
-        URLClassLoader compileClassLoader = JKubeProjectUtil.getClassLoader(javaProject);
-
-        // When
-        Properties properties = SpringBootUtil.getSpringBootApplicationProperties(compileClassLoader);
-
-        // Then
-        assertThat(properties)
-            .containsEntry("server.port", "8081");
-    }
-
-
-    @Test
-    void testYamlToPropertiesParsing() {
-
-        Properties props = YamlUtil.getPropertiesFromYamlResource(getClass().getResource("/util/springboot/test-application.yml"));
-        assertThat(props).isNotEmpty()
-            .contains(
-                entry("management.port", "8081"),
-                entry("spring.datasource.url", "jdbc:mysql://127.0.0.1:3306"),
-                entry("example.nested.items[0].value", "value0"),
-                entry("example.nested.items[1].value", "value1"),
-                entry("example.nested.items[2].elements[0].element[0].subelement", "sub0"),
-                entry("example.nested.items[2].elements[0].element[1].subelement", "sub1"),
-                entry("example.1", "integerKeyElement"));
+    void whenMultipleNativeExecutablesInStandardMavenBuildDirectory_thenThrowException() throws IOException {
+      // Given
+      File nativeArtifactFile = Files.createFile(tempDir.resolve("sample")).toFile();
+      assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
+      File nativeArtifactFile2 = Files.createFile(tempDir.resolve("sample2")).toFile();
+      assertThat(nativeArtifactFile2.setExecutable(true)).isTrue();
+      // When + Then
+      assertThatIllegalStateException()
+        .isThrownBy(() -> SpringBootUtil.findNativeArtifactFile(javaProject))
+        .withMessage("More than one native executable file found in " + tempDir.toFile().getAbsolutePath());
     }
 
     @Test
-    void testInvalidFileThrowsException() {
-        URL resource = SpringBootUtil.class.getResource("/util/springboot/invalid-application.yml");
-        assertThrows(IllegalStateException.class, () -> YamlUtil.getPropertiesFromYamlResource(resource));
+    void whenMultipleExecutableFilesButOnlyOneNativeExecutableInStandardMavenBuildDirectory_thenReturnNativeArtifact() throws IOException {
+      // Given
+      File nativeArtifactFile = Files.createFile(tempDir.resolve("sample")).toFile();
+      assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
+      for (String fileName : new String[]{"libjava.so", "libjvm.so"}) {
+        File nonNativeArtifactFile = Files.createFile(tempDir.resolve(fileName)).toFile();
+        assertThat(nonNativeArtifactFile.setExecutable(true)).isTrue();
+      }
+      // When
+      File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
+      // Then
+      assertThat(nativeArtifactFound).hasName("sample");
+    }
+  }
+
+  @Nested
+  class HasSpringWebFluxDependency {
+    @Test
+    void whenWebFluxDependencyPresent_thenReturnTrue() {
+      // Given
+      JavaProject javaProject = JavaProject.builder().dependency(Dependency.builder()
+          .groupId("org.springframework.boot")
+          .artifactId("spring-boot-starter-webflux")
+          .build())
+        .build();
+      // When + Then
+      assertThat(SpringBootUtil.hasSpringWebFluxDependency(javaProject)).isTrue();
     }
 
     @Test
-    void testNonExistentYamlToPropertiesParsing() {
-        Properties props = YamlUtil.getPropertiesFromYamlResource(getClass().getResource("/this-file-does-not-exist"));
-        assertThat(props).isNotNull().isEmpty();
+    void whenNoWebFluxDependencyPresent_thenReturnFalse() {
+      // Given
+      JavaProject javaProject = JavaProject.builder().dependency(Dependency.builder()
+        .groupId("org.springframework.boot")
+        .artifactId("spring-boot-starter-web")
+        .build()).build();
+      // When + Then
+      assertThat(SpringBootUtil.hasSpringWebFluxDependency(javaProject)).isFalse();
     }
+  }
 
-    @Test
-    void getSpringBootApplicationProperties_withMultipleProfilesAndDefault(@TempDir File tempDir) throws IOException {
-        URLClassLoader cl = createClassLoader(tempDir, "/util/springboot/test-application-with-multiple-profiles.yml");
-        Properties props = SpringBootUtil.getSpringBootApplicationProperties(cl);
-        assertThat(props).isNotEmpty()
-            .contains(
-                entry("spring.application.name", "spring-boot-k8-recipes"),
-                entry("management.endpoints.enabled-by-default", "false"),
-                entry("management.endpoint.health.enabled", "true"))
-            .doesNotContainEntry("cloud.kubernetes.reload.enabled", null);
-    }
-
-    @Test
-    void getSpringBootApplicationProperties_withMultipleProfilesAndSpecificProfile(@TempDir File tempDir) throws IOException {
-        URLClassLoader cl = createClassLoader(tempDir, "/util/springboot/test-application-with-multiple-profiles.yml");
-        Properties props = SpringBootUtil.getSpringBootApplicationProperties("kubernetes", cl);
-        assertThat(props)
-            .containsEntry("cloud.kubernetes.reload.enabled", "true")
-            .doesNotContain(
-                entry("cloud.kubernetes.reload.enabled", null),
-                entry("spring.application.name", null));
-    }
-
-    @Test
-    void getSpringBootPluginConfiguration_whenNothingPresent_thenReturnsEmptyMap() {
-        // Given
-        JavaProject javaProject = JavaProject.builder().build();
-
-        // When
-        Map<String, Object> configuration = SpringBootUtil.getSpringBootPluginConfiguration(javaProject);
-
-        // Then
-        assertThat(configuration).isEmpty();
-    }
-
-    @Test
-    void getSpringBootPluginConfiguration_whenSpringBootMavenPluginPresent_thenReturnsPluginConfiguration() {
-        // Given
-        JavaProject javaProject = JavaProject.builder()
-            .plugin(Plugin.builder()
-                .groupId("org.springframework.boot")
-                .artifactId("spring-boot-maven-plugin")
-                .configuration(Collections.singletonMap("layout", "ZIP"))
-                .build())
-            .build();
-
-        // When
-        Map<String, Object> configuration = SpringBootUtil.getSpringBootPluginConfiguration(javaProject);
-
-        // Then
-        assertThat(configuration).isNotNull().containsEntry("layout", "ZIP");
-    }
-
-    @Test
-    void getSpringBootPluginConfiguration_whenSpringBootGradlePluginPresent_thenReturnsPluginConfiguration() {
-        // Given
-        JavaProject javaProject = JavaProject.builder()
-            .plugin(Plugin.builder()
-                .groupId("org.springframework.boot")
-                .artifactId("org.springframework.boot.gradle.plugin")
-                .configuration(Collections.singletonMap("mainClass", "com.example.ExampleApplication"))
-                .build())
-            .build();
-
-        // When
-        Map<String, Object> configuration = SpringBootUtil.getSpringBootPluginConfiguration(javaProject);
-
-        // Then
-        assertThat(configuration).isNotNull().containsEntry("mainClass", "com.example.ExampleApplication");
-    }
-
-    @Test
-    void isSpringBootRepackage_whenPluginHasRepackageExecution_thenReturnTrue() {
-        // Given
-        JavaProject javaProject = JavaProject.builder()
-            .plugin(Plugin.builder()
-                .groupId("org.springframework.boot")
-                .artifactId("spring-boot-maven-plugin")
-                .executions(Collections.singletonList("repackage"))
-                .build())
-            .build();
-
-        // When
-        boolean result = SpringBootUtil.isSpringBootRepackage(javaProject);
-
-        // Then
-        assertThat(result).isTrue();
-    }
-
-    @Test
-    void isSpringBootRepackage_whenNoExecution_thenReturnFalse() {
-        // Given
-        JavaProject javaProject = JavaProject.builder()
-            .plugin(Plugin.builder()
-                .groupId("org.springframework.boot")
-                .artifactId("spring-boot-maven-plugin")
-                .executions(Collections.emptyList())
-                .build())
-            .build();
-
-        // When
-        boolean result = SpringBootUtil.isSpringBootRepackage(javaProject);
-
-        // Then
-        assertThat(result).isFalse();
-    }
-
-    @Test
-    void getNativePlugin_whenNoNativePluginPresent_thenReturnNull() {
-        assertThat(SpringBootUtil.getNativePlugin(JavaProject.builder().build())).isNull();
-    }
-
-    @Test
-    void getNativePlugin_whenMavenNativePluginPresent_thenReturnPlugin() {
-        // Given
-        JavaProject javaProject = JavaProject.builder()
-            .plugin(Plugin.builder()
-                .groupId("org.graalvm.buildtools")
-                .artifactId("native-maven-plugin")
-                .build())
-            .build();
-
-        // When
-        Plugin plugin = SpringBootUtil.getNativePlugin(javaProject);
-
-        // Then
-        assertThat(plugin).isNotNull();
-    }
-
-    @Test
-    void getNativePlugin_whenGradleNativePluginPresent_thenReturnPlugin() {
-        // Given
-        JavaProject javaProject = JavaProject.builder()
-            .plugin(Plugin.builder()
-                .groupId("org.graalvm.buildtools.native")
-                .artifactId("org.graalvm.buildtools.native.gradle.plugin")
-                .build())
-            .build();
-
-        // When
-        Plugin plugin = SpringBootUtil.getNativePlugin(javaProject);
-
-        // Then
-        assertThat(plugin).isNotNull();
-    }
-
-    @Test
-    void findNativeArtifactFile_whenNativeExecutableNotFound_thenReturnNull(@TempDir File temporaryFolder) {
-        // Given
-        JavaProject javaProject = JavaProject.builder()
-            .artifactId("sample")
-            .buildDirectory(temporaryFolder)
-            .plugin(Plugin.builder()
-                .groupId("org.graalvm.buildtools")
-                .artifactId("native-maven-plugin")
-                .build())
-            .build();
-
-        // When
-        File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
-
-        // Then
-        assertThat(nativeArtifactFound).isNull();
-    }
-
-    @Test
-    void findNativeArtifactFile_whenNativeExecutableInStandardMavenBuildDirectory_thenReturnNativeArtifact(@TempDir File temporaryFolder) throws IOException {
-        // Given
-        File nativeArtifactFile = Files.createFile(temporaryFolder.toPath().resolve("sample")).toFile();
-        assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
-        JavaProject javaProject = JavaProject.builder()
-            .artifactId("sample")
-            .buildDirectory(temporaryFolder)
-            .plugin(Plugin.builder()
-                .groupId("org.graalvm.buildtools")
-                .artifactId("native-maven-plugin")
-                .build())
-            .build();
-
-        // When
-        File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
-
-        // Then
-        assertThat(nativeArtifactFound).hasName("sample");
-    }
-
-    @Test
-    void findNativeArtifactFile_whenMoreThanOneNativeExecutableInStandardMavenBuildDirectory_thenThrowException(@TempDir File temporaryFolder) throws IOException {
-        // Given
-        File nativeArtifactFile = Files.createFile(temporaryFolder.toPath().resolve("sample")).toFile();
-        assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
-        File nativeArtifactFile2 = Files.createFile(temporaryFolder.toPath().resolve("sample2")).toFile();
-        assertThat(nativeArtifactFile2.setExecutable(true)).isTrue();
-        JavaProject javaProject = JavaProject.builder()
-            .artifactId("sample")
-            .buildDirectory(temporaryFolder)
-            .plugin(Plugin.builder()
-                .groupId("org.graalvm.buildtools")
-                .artifactId("native-maven-plugin")
-                .build())
-            .build();
-
-        // When + Then
-        assertThatIllegalStateException()
-            .isThrownBy(() -> SpringBootUtil.findNativeArtifactFile(javaProject))
-            .withMessage("More than one native executable file found in " + temporaryFolder.getAbsolutePath());
-    }
-
-    @Test
-    void findNativeArtifactFile_whenNativeExecutableInStandardMavenBuildDirectoryAndImageNameOverridden_thenReturnNativeArtifact(@TempDir File temporaryFolder) throws IOException {
-        // Given
-        File nativeArtifactFile = Files.createFile(temporaryFolder.toPath().resolve("custom-native-name")).toFile();
-        assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
-        JavaProject javaProject = JavaProject.builder()
-            .artifactId("sample")
-            .buildDirectory(temporaryFolder)
-            .plugin(Plugin.builder()
-                .groupId("org.graalvm.buildtools")
-                .artifactId("native-maven-plugin")
-                .build())
-            .build();
-
-        // When
-        File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
-
-        // Then
-        assertThat(nativeArtifactFound).hasName("custom-native-name");
-    }
-
-    @Test
-    void findNativeArtifactFile_whenNativeExecutableInStandardGradleNativeDirectory_thenReturnNativeArtifact(@TempDir File temporaryFolder) throws IOException {
-        // Given
-        Files.createDirectories(temporaryFolder.toPath().resolve("native").resolve("nativeCompile"));
-        File nativeArtifactFile = Files.createFile(temporaryFolder.toPath().resolve("native").resolve("nativeCompile").resolve("sample")).toFile();
-        assertThat(nativeArtifactFile.setExecutable(true)).isTrue();
-        JavaProject javaProject = JavaProject.builder()
-            .artifactId("sample")
-            .buildDirectory(temporaryFolder)
-            .plugin(Plugin.builder()
-                .groupId("org.graalvm.buildtools.native")
-                .artifactId("org.graalvm.buildtools.native.gradle.plugin")
-                .build())
-            .build();
-
-        // When
-        File nativeArtifactFound = SpringBootUtil.findNativeArtifactFile(javaProject);
-
-        // Then
-        assertThat(nativeArtifactFound).hasName("sample");
-    }
-
-    @Test
-    void hasSpringWebFluxDependency_whenWebFluxDependencyPresent_thenReturnTrue() {
-        // Given
-        JavaProject javaProject = JavaProject.builder().dependency(Dependency.builder()
-            .groupId("org.springframework.boot")
-            .artifactId("spring-boot-starter-webflux")
-            .build())
-          .build();
-        // When + Then
-        assertThat(SpringBootUtil.hasSpringWebFluxDependency(javaProject)).isTrue();
-    }
-
-    @Test
-    void hasSpringWebFluxDependency_whenNoWebFluxDependencyPresent_thenReturnFalse() {
-        // Given
-        JavaProject javaProject = JavaProject.builder().dependency(Dependency.builder()
-            .groupId("org.springframework.boot")
-            .artifactId("spring-boot-starter-web")
-            .build()).build();
-        // When + Then
-        assertThat(SpringBootUtil.hasSpringWebFluxDependency(javaProject)).isFalse();
-    }
-
-    static URLClassLoader createClassLoader(File temporaryFolder, String resource) throws IOException {
-        File applicationProp =  new File(Objects.requireNonNull(SpringBootUtilTest.class.getResource(resource)).getPath());
-        File classesInTarget = new File(new File(temporaryFolder, "target"), "classes");
-        FileUtil.createDirectory(classesInTarget);
-        File applicationPropertiesInsideTarget = new File(classesInTarget, "application." +
-          applicationProp.getName().substring(applicationProp.getName().lastIndexOf(".") + 1));
-        FileUtils.copyFile(applicationProp, applicationPropertiesInsideTarget);
-        return ClassUtil.createClassLoader(Arrays.asList(classesInTarget.getAbsolutePath(), applicationProp.getAbsolutePath()));
-    }
+  static URLClassLoader createClassLoader(File temporaryFolder, String resource) throws IOException {
+    File applicationProp = new File(Objects.requireNonNull(SpringBootUtilTest.class.getResource(resource)).getPath());
+    File classesInTarget = new File(new File(temporaryFolder, "target"), "classes");
+    FileUtil.createDirectory(classesInTarget);
+    File applicationPropertiesInsideTarget = new File(classesInTarget, "application." +
+      applicationProp.getName().substring(applicationProp.getName().lastIndexOf(".") + 1));
+    FileUtils.copyFile(applicationProp, applicationPropertiesInsideTarget);
+    return ClassUtil.createClassLoader(Arrays.asList(classesInTarget.getAbsolutePath(), applicationProp.getAbsolutePath()));
+  }
 }

